test/test-condvar.c

```diff
@@ -120,64 +139,76 @@ TEST_IMPL(condvar_2) {
   ASSERT(0 == uv_thread_join(&thread));
   uv_mutex_destroy(&wc.mutex);
   uv_cond_destroy(&wc.cond);
+  uv_sem_destroy(&wc.sem_signaled);
+  uv_sem_destroy(&wc.sem_waiting);
 
   return 0;
 }
 
-
+/* 1. Wait on a signal (hopefully not timeout, else we'll hang).
+ * 2. Ensure that the signaler finished. */
 static int condvar_timedwait(worker_config* c, const volatile int* flag) {
   int r;
 
   r = 0;
 
   uv_mutex_lock(&c->mutex);
-  if (c->wait_delay)
-    uv_sleep(c->wait_delay);
-  while (*flag == 0) {
-    r = uv_cond_timedwait(&c->cond, &c->mutex, (uint64_t)(150 * 1e6));
-    ASSERT(r == 0 || r == UV_ETIMEDOUT);
-    if (r == UV_ETIMEDOUT)
-      break;
-  }
+
+  /* Tell signal'er that I am waiting. */
+  uv_sem_post(&c->sem_waiting);
+
+  /* Wait until I get a non-spurious signal. */
+  do {
+    r = uv_cond_timedwait(&c->cond, &c->mutex, (uint64_t)(1 * 1e9)); /* 1 s */
+    ASSERT(r == 0); /* Should not time out. */
+  } while (*flag == 0);
+  ASSERT(*flag == 1);
+
   uv_mutex_unlock(&c->mutex);
 
+  /* Wait for my signal'er to finish. */
+  uv_sem_wait(&c->sem_signaled);
   return r;
 }
 
-/* Test that uv_cond_timedwait will return early when cond is signaled. */
+/* uv_cond_timedwait: One thread signals, the other waits. */
 TEST_IMPL(condvar_3) {
   uv_thread_t thread;
   worker_config wc;
 
   memset(&wc, 0, sizeof(wc));
-  wc.signal_delay = 100;
   wc.signal_cond = condvar_signal;
   wc.wait_cond = condvar_timedwait;
 
+  ASSERT(0 == uv_sem_init(&wc.sem_waiting, 0));
+  ASSERT(0 == uv_sem_init(&wc.sem_signaled, 0));
   ASSERT(0 == uv_cond_init(&wc.cond));
   ASSERT(0 == uv_mutex_init(&wc.mutex));
   ASSERT(0 == uv_thread_create(&thread, worker, &wc));
 
-  ASSERT(0 == wc.wait_cond(&wc, &wc.posted_1));
+  wc.wait_cond(&wc, &wc.posted_1);
   wc.signal_cond(&wc, &wc.posted_2);
 
   ASSERT(0 == uv_thread_join(&thread));
   uv_mutex_destroy(&wc.mutex);
   uv_cond_destroy(&wc.cond);
+  uv_sem_destroy(&wc.sem_signaled);
+  uv_sem_destroy(&wc.sem_waiting);
 
   return 0;
 }
 
-
+/* uv_cond_timedwait: One thread broadcasts, the other waits. */
 TEST_IMPL(condvar_4) {
```

Comment:
Maybe I'm missing something, but this test looks exactly the same as the `condvar_3` test (it's not broadcasting)